internal/controller/maintenancemode_controller.go

```diff
@@ -0,0 +1,202 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	maintenanceModeFinalizer = "ocs-client-operator.ocs.openshift.io/maintenance-mode"
+	ramenReplicationIDLabel  = "ramendr.openshift.io/replicationID"
+	MaintenanceModeCRDName   = "maintenancemodes.ramendr.openshift.io"
+)
+
+// MaintenanceModeReconciler reconciles a ClusterVersion object
+type MaintenanceModeReconciler struct {
+	client.Client
+	OperatorNamespace string
+	Scheme            *runtime.Scheme
+
+	log logr.Logger
+	ctx context.Context
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MaintenanceModeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	generationChangePredicate := predicate.GenerationChangedPredicate{}
+	bldr := ctrl.NewControllerManagedBy(mgr).
+		For(&ramenv1alpha1.MaintenanceMode{}, builder.WithPredicates(generationChangePredicate)).
+		Watches(&storagev1.StorageClass{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(generationChangePredicate)).
+		Watches(&v1alpha1.StorageClient{}, &handler.EnqueueRequestForObject{}, builder.WithPredicates(generationChangePredicate))
+
+	return bldr.Complete(r)
+}
+
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes,verbs=get;list;update;create;watch;delete
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=ramendr.openshift.io,resources=maintenancemodes/finalizers,verbs=update
+//+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclients,verbs=get;list;watch
+//+kubebuilder:rbac:groups=storage.k8s.io,resources=storageclasses,verbs=get;list;watch
+
+func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx, "MaintenanceMode", req)
+	r.log.Info("Reconciling MaintenanceMode")
+
+	maintenanceMode := &ramenv1alpha1.MaintenanceMode{}
+	maintenanceMode.Name = req.Name
+	if err := r.get(maintenanceMode); err != nil {
+		if kerrors.IsNotFound(err) {
+			r.log.Info("Maintenance Mode resource not found. Ignoring since object might be deleted.")
+			return reconcile.Result{}, nil
+		}
+		r.log.Error(err, "failed to get the Maintenance Mode")
+		return reconcile.Result{}, err
+	}
+
+	// find the storageClass with the targetID
+	storageClassList := &storagev1.StorageClassList{}
+	err := r.list(storageClassList, client.MatchingLabels{ramenReplicationIDLabel: maintenanceMode.Spec.TargetID})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if len(storageClassList.Items) != 1 {
+		return ctrl.Result{}, fmt.Errorf("failed to find storageClass for maintenance mode")
+	}
+
+	storageClass := &storageClassList.Items[0]
```

Comment:
Why the first storage class? how do you know that the first one belongs to the/a proper client?

Reply:
We are finding out the storageClass via the replicationID, which is uniqueID added to the storageClass required for ramen. There will be only one such storageClass, I have a check just above to fail if we find multiple storageClasses with the same replicationID

Reply:
I am not sure replicationID is unique to a storage class. Who selects it and sets it up?

Reply:
Both replicationID and storageID are unique and they are set from the provider server.
replicationID is the hash of storageClass name (so that it is same across cluster, this is a requirement from ramen) and storageID is also unique for each storageclass